### PR TITLE
[MC-1129] Orca Security plugin - 11 actions and 1 trigger

### PR DESCRIPTION
**Pull request by dsliwinski-r7**
### Description

Describe the proposed changes:

  - Create Orca Security plugin

## PR Requirements

Developers, verify you have completed the following items by checking them off:

#### Unit Tests

Review our documentation on [generating](https://docs.rapid7.com/insightconnect/unit-test-generation) and [writing](https://docs.rapid7.com/insightconnect/unit-test-primer) plugin unit tests

- [ ] Unit tests written for any new or updated code

#### In-Product Tests

If you are an InsightConnect customer or have access to an InsightConnect instance, the following in-product tests should be done:

- [ ] Screenshot of job output with the plugin changes
- [ ] Screenshot of the changed connection, actions, or triggers input within the InsightConnect workflow builder

### Style

Review the [style guide](https://docs.rapid7.com/insightconnect/style-guide/)

- [ ] For dependencies, pin [OS package](https://docs.rapid7.com/insightconnect/style-guide/#dockerfile) and [Python package](https://docs.rapid7.com/insightconnect/style-guide/#requirements.txt) versions
- [ ] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [ ] For size, use the [slim SDK images](https://docs.rapid7.com/insightconnect/sdk-guide/#sdk-guide) when possible: ``komand/python-3-37-slim-plugin`` and ``komand/python-3-37-plugin``
- [ ] For error handling, use of [PluginException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations/#plugin-exceptions) and [ConnectionTestException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations#connection-exceptions)
- [ ] For logging, use [self.logger](https://docs.rapid7.com/insightconnect/sdk-guide/#logging)
- [ ] For docs, use [changelog style](https://docs.rapid7.com/insightconnect/style-guide/#changelog)
- [ ] For docs, validate markdown with ``make validate`` which calls ``mdl`` to lint ``help.md``

### Functional Checklist
- [ ] Work fully completed
- [ ] Functional
  - [ ] Any new actions/triggers include JSON [test files](https://docs.rapid7.com/insightconnect/style-guide/#tests) in the `tests/` directory created with `icon-plugin run -c sample $action > tests/$action.json`
  - [ ] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [ ] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [ ] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -T all --debug --jq` (use PR format at end)
  - [ ] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -R all --debug --jq` (use PR format at end)

### Assessment

[assessments.md](https://github.com/rapid7/insightconnect-plugins/files/9372626/assessments.md)

### CODE REVIEW

**Comment by dsliwinski-r7** [comment]
PluginValidator fails due to missing icon.png and extension.png files.

**Review by mgrandys-r7 — APPROVED** [review]

**Review by dmandyna-r7 — APPROVED** [review]

File: plugins/orca_security/unit_test/parameters/get_asset_by_id.json.resp [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+          "cluster_type": "asg",
+          "type": "vm",
+          "group_unique_id": "group-12345",
+          "tags_value_list": [
+            "ECSAutoScalingGroup"
+          ],
+          "vm_asset_unique_ids": [
+            "vm_12345"
+          ],
+          "cloud_account_id": "44d88612-fea8-a8f3-6de8-2e1278abb02f",
```
Comment (igorski-r7) [inline plugins/orca_security/unit_test/parameters/get_asset_by_id.json.resp:22]: I believe all the values are faked in `.json.resp` files 😄

File: plugins/orca_security/unit_test/test_add_user.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        actual = self.action.run(
+            {
+                Input.INVITE_USER_EMAIL: email,
+                Input.ROLE_ID: role_id,
+                Input.ALL_CLOUD_ACCOUNTS: all_cloud_accounts,
+                Input.CLOUD_ACCOUNTS: cloud_accounts,
+                Input.SHOULD_SEND_EMAIL: should_send_email,
+            }
+        )
+        self.assertEqual(actual, expected)
```
Comment (igorski-r7) [inline plugins/orca_security/unit_test/test_add_user.py:34]: Kindly, could you please add tests including all the `exceptions` raised for that action? 😄

File: plugins/orca_security/unit_test/test_delete_user.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+@patch("requests.request", side_effect=Util.mocked_requests)
+class TestDeleteUser(TestCase):
+    @classmethod
+    def setUpClass(cls) -> None:
+        cls.action = Util.default_connector(DeleteUser())
+
+    @parameterized.expand(Util.load_parameters("delete_user").get("parameters"))
+    def test_delete_user(self, mock_request, name, email, expected):
+        actual = self.action.run({Input.DELETE_INVITE_EMAIL: email})
+        self.assertEqual(actual, expected)
```
Comment (igorski-r7) [inline plugins/orca_security/unit_test/test_delete_user.py:24]: Same here 😄

File: plugins/orca_security/unit_test/test_download_malicious_file.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        actual = self.action.run({Input.ALERT_ID: alert_id})
+        self.assertEqual(actual, expected)
+
+    @parameterized.expand(Util.load_parameters("download_malicious_file_bad").get("parameters"))
+    def test_download_malicious_file_bad(self, mock_request, mock_get, name, alert_id, cause, assistance, data):
+        with self.assertRaises(PluginException) as e:
+            self.action.run({Input.ALERT_ID: alert_id})
+        self.assertEqual(e.exception.cause, cause)
+        self.assertEqual(e.exception.assistance, assistance)
+        self.assertEqual(e.exception.data, data)
```
Comment (igorski-r7) [inline plugins/orca_security/unit_test/test_download_malicious_file.py:33]: I would change `e` to `error` here, but it's up to you 😄

File: plugins/orca_security/unit_test/test_get_alerts.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+@patch("requests.request", side_effect=Util.mocked_requests)
+class TestGetAlerts(TestCase):
+    @classmethod
+    def setUpClass(cls) -> None:
+        cls.action = Util.default_connector(GetAlerts())
+
+    @parameterized.expand(Util.load_parameters("get_alerts").get("parameters"))
+    def test_get_alerts(self, mock_request, name, filters, limit, expected):
+        actual = self.action.run({Input.FILTERS: filters, Input.LIMIT: limit})
+        self.assertEqual(actual, expected)
```
Comment (igorski-r7) [inline plugins/orca_security/unit_test/test_get_alerts.py:24]: Same here. Kindly, could you please add all the `exceptions` test for that action? 😄

File: plugins/orca_security/unit_test/test_get_assets.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+                Input.ASSET_STATE: asset_state,
+                Input.ASSET_LABELS: asset_label,
+                Input.STATE_SCORE: score,
+                Input.STATE_SEVERITY: severity,
+                Input.INTERNET_FACING: internet_facing,
+                Input.COMPUTE_REGIONS: region,
+                Input.COMPUTE_VPCS: vpc,
+            }
+        )
+        self.assertEqual(actual, expected)
```
Comment (igorski-r7) [inline plugins/orca_security/unit_test/test_get_assets.py:52]: Same here

File: plugins/orca_security/unit_test/test_get_users.py [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+@patch("requests.request", side_effect=Util.mocked_requests)
+class TestGetUsers(TestCase):
+    @classmethod
+    def setUpClass(cls) -> None:
+        cls.action = Util.default_connector(GetUsers())
+
+    @parameterized.expand(Util.load_parameters("get_users").get("parameters"))
+    def test_get_users(self, mock_request, name, expected):
+        actual = self.action.run()
+        self.assertEqual(actual, expected)
```
Comment (igorski-r7) [inline plugins/orca_security/unit_test/test_get_users.py:24]: Same here

**Review by igorski-r7 — CHANGES_REQUESTED** [review]
Looks great! 😄 Left some small comments.

File: plugins/orca_security/unit_test/parameters/get_asset_by_id.json.resp (hunk 1 above)
Comment (dsliwinski-r7) [reply plugins/orca_security/unit_test/parameters/get_asset_by_id.json.resp:22]: Yes, all sensitive data is not real :)

File: plugins/orca_security/unit_test/test_get_users.py (hunk 7 above)
Comment (dsliwinski-r7) [reply plugins/orca_security/unit_test/test_get_users.py:24]: There is no exceptions for this action that I could test :)

**Review by igorski-r7 — APPROVED** [review]

File: plugins/orca_security/icon_orca_security/triggers/new_alert/trigger.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+        while True:
+            dsl_filter = {"dsl_filter": json.dumps({"filter": filters})}
+            filters[0] = {"field": "state.created_at", "range": {"gt": datetime.now().isoformat()}}
+            alerts = self.connection.api.query_alerts(dsl_filter).get("data", [])
+            for alert in alerts:
+                self.send({Output.ALERT: alert})
+            time.sleep(interval)
+
+    @staticmethod
+    def validate_filters(filters: list) -> list:
```
Comment (jkaluza-rapid7) [inline plugins/orca_security/icon_orca_security/triggers/new_alert/trigger.py:31]: id move that and other static methods to a separate file in utils

**Review by jkaluza-rapid7 — CHANGES_REQUESTED** [review]

File: plugins/orca_security/icon_orca_security/triggers/new_alert/trigger.py (hunk 8 above)
Comment (mgrandys-r7) [reply plugins/orca_security/icon_orca_security/triggers/new_alert/trigger.py]: @jkaluza-rapid7 Done :)

**Review by jkaluza-rapid7 — APPROVED** [review]

**Pull request merged**